### ISSUE TITLE
UTF-8 code unit operations

### ISSUE THREAD

**Issue by ollef** [Bodigrim/text-rope#3]
Hey!

How do you feel about e.g.

```haskell
utf8Length :: Rope -> Word
utf8SplitAt :: Word -> Rope -> Maybe (Rope, Rope)
```

for working with UTF-8 code units? The use case is to convert a UTF-16 position obtained from the LSP library to UTF-8 code units (which my compiler uses internally).

For my use case it's probably fine for these to work without a cached measure if that would negatively impact performance elsewhere. `utf8Length` would e.g. be a fold over the chunks.

These functions might have to be predicated on text version >= 2.0 (or be quite a bit more inefficient when using UTF-16 text). Thoughts?

I could try to contribute these (unless there's a reason not to have them).

Cheers!

**Comment by Bodigrim** [comment]
> The use case is to convert a UTF-16 position obtained from the LSP library to UTF-8 code units (which my compiler uses internally).

Isn't https://hackage.haskell.org/package/text-rope-0.2/docs/Data-Text-Utf16-Rope-Mixed.html up for the purpose? I think `lsp` is migrating to use it internally as well in https://github.com/haskell/lsp/commit/847c2390d00d67f2e999d387a3eea7f602eb5ab4. @michaelpj any plans to make a release of `lsp` soon?

**Comment by ollef** [comment]
> Isn't https://hackage.haskell.org/package/text-rope-0.2/docs/Data-Text-Utf16-Rope-Mixed.html up for the purpose?

As far as I can tell that interface is for working with indices in code points, which is independent of encoding, and UTF-16 code units. You could achieve what I want by converting to `Text` and then splitting or taking the UTF-8 length of the text, but that seems to defeat the purpose of using ropes (splitting and taking the length is constant time but converting to text is linear). Is there a more clever way to do it that I'm not seeing?

Another option would be to make `Mixed` also track UTF-8 lengths, making it fast for code points, UTF-16 code units _and_ UTF-8 code units. It might be worth at least trying that to see what the performance impact is.

**Comment by Bodigrim** [comment]
I'm not opposed to such extension, but I'd prefer a new interface to be consistent with the rest of the library: e. g., UTF-8-based splitting and indexing should have the same performance complexity as UTF-16- and char-based. This is likely to entail a significant amount of work. 

It might be easier to provide generic combinators like the following:
```haskell
pseudoLength :: Monoid m => (Text -> m) -> Rope -> m
pseudoSplitAt :: (Monoid m, Ord m) => (m -> Text -> Maybe (Text, Text)) -> m -> Rope -> Maybe (Rope, Rope)
```
(or should it be `TextLines` instead of `Text`?..)

**Comment by Bodigrim** [comment]
Closed by #4, I suppose.
